app/assets/src/components/views/PipelineViz.jsx

```diff
@@ -49,154 +54,324 @@ class PipelineViz extends React.Component {
+        step.in.forEach(inTarget => {
+          targets[inTarget].forEach(fileName => {
+            const filePathSections =
+              inTarget in stageData.given_targets
+                ? [stageData.given_targets[inTarget].s3_dir]
+                : [stageData.output_dir_s3, this.pipelineVersion];
+            filePathSections.push(fileName);
+
+            const filePath = this.createFilePath(filePathSections);
+            remainingOutputFiles.delete(filePath);
+
+            infoPerFilePath.push({
+              filePath: filePath,
+              fileName: fileName,
+              to: {
+                stageIndex: stageIndex,
+                stepIndex: stepIndex,
+              },
+              ...(filePathToOutputtingStep[filePath] || {}),
+            });
+          });
+        });
+      });
+    });
+
+    // Include output files that aren't input to any steps.
+    Array.from(remainingOutputFiles).forEach(filePath => {
+      infoPerFilePath.push({
+        filePath: filePath,
+        ...filePathToOutputtingStep[filePath],
+      });
+    });
+
+    return infoPerFilePath;
   }
 
-  stageResultsWithModifiedStepNames() {
-    // Strips 'PipelineStep[Run/Generate]' from front of each step name.
-    // TODO(ezhong): Consider adding 'name' field to dag_json later.
+  createStageStepData() {
     const { stageResults } = this.props;
-    const stagesWithModifiedNames = mapValues(stageData => {
-      const modifiedStageData = Object.assign({}, stageData);
-      modifiedStageData.steps = modifiedStageData.steps.map(step => {
+
+    const stages = this.stageNames.map(stageName => {
+      const stageData = stageResults.stages[stageName];
+      const steps = stageData.steps.map(step => {
         return {
-          ...step,
-          class: step.class.replace(/^(PipelineStep(Run|Generate)?)/, ""),
+          name: this.getModifiedStepName(step.class),
+          inputEdges: [],
+          outputEdges: [],
         };
       });
-      return modifiedStageData;
-    }, stageResults.stages);
-    return stagesWithModifiedNames;
+      return {
+        stageName: stageName,
+        jobStatus: stageData.job_status,
+        steps: steps,
+      };
+    });
+
+    return stages;
   }
 
-  handleMouseWheelZoom = e => {
-    const { zoomChangeInterval } = this.props;
-    const zoomChange = (e.deltaY < 0 ? 1 : -1) * zoomChangeInterval;
-    this.setState({ zoom: this.state.zoom + zoomChange });
-  };
+  populateStageStepDataEdges() {
+    this.edgeFileData.forEach((edgeData, edgeIndex) => {
+      const { from, to } = edgeData;
+      from && this.getStepDataAt(from).outputEdges.push(edgeIndex);
+      to && this.getStepDataAt(to).inputEdges.push(edgeIndex);
+    });
+  }
+
+  getEdgesBetweenNodes(graph, fromId, toId, type) {
+    const allEdges = graph.getEdgesBetweenNodes(fromId, toId);
+    if (type) {
+      const edgeTypeRegex = new RegExp(`-${type}$`, "g");
+      return allEdges.filter(edgeId => edgeId.match(edgeTypeRegex));
```

Comment:
Have you thought of using a graph data structure to store everything? That would make this kind of retrieval O(1) instead of having to loop through all edges.
We can also define a map indexed by `from` and `to` nodes to make this a but more efficient.

Reply:
Any comments on this?

Reply:
I'm trying to refactor it currently based off the edgeIds I've created in `generateEdgeData`, which should make it more efficient

Reply:
Looking at the code again, I realized that I'm looking through connected edges to a node multiple times for one hover action (ie once in my call to `getEdgesBetweenNodes` and then again when looking through the results of `this.getEdgeInfoFor` — I've moved all the logic into looping through the result of `this.getEdgeInfoFor`
https://github.com/chanzuckerberg/idseq-web/pull/2376/commits/6cb2059cc181abf5855f3dada9c495067318fb85#diff-749f700b60f6b84cf7f1ce76d874bff2R284